src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py

```diff
@@ -368,36 +398,89 @@ def _predict_tokens(
                     for i in start_token_ids + end_token_ids:
                         token_id_to_word_id[i] = first_token_id
+                        word_id_to_token_ids[first_token_id].append(i)
                     word_id_to_text[first_token_id] = candidate_text
                 # if concatenated tokens wout hyphen are in dictionary
                 elif internal_dictionary.is_in(text=candidate_text_no_hyphen, strip_punct=False):
                     for i in start_token_ids + end_token_ids:
                         token_id_to_word_id[i] = first_token_id
+                        word_id_to_token_ids[first_token_id].append(i)
                     word_id_to_text[first_token_id] = candidate_text_no_hyphen
                 # if concatenated tokens wout hyphen *AND* right-strip punct are in dict..
                 elif internal_dictionary.is_in(text=candidate_text_strip_punct_no_hyphen,
                                                strip_punct=False):
                     for i in start_token_ids + end_token_ids_strip_punct:
                         token_id_to_word_id[i] = first_token_id
+                        word_id_to_token_ids[first_token_id].append(i)
                     word_id_to_text[first_token_id] = candidate_text_strip_punct_no_hyphen
                     for i in end_token_ids:
-                        if i in punct_strip_candidate_token_ids:
+                        if i in punct_r_strip_candidate_token_ids:
                             token_id_to_word_id[i] = i
+                            word_id_to_token_ids[i].append(i)
                             word_id_to_text[i] = document.tokens[i].text
                 # if concatenated tokens are *NOT* in dictionary, default behavior is
                 # to concatenate anyways, keeping hyphen, and stripping punctuation as
                 # separate tokens
                 else:
                     for i in start_token_ids + end_token_ids_strip_punct:
                         token_id_to_word_id[i] = first_token_id
+                        word_id_to_token_ids[first_token_id].append(i)
                     word_id_to_text[first_token_id] = candidate_text_strip_punct
                     for i in end_token_ids:
-                        if i in punct_strip_candidate_token_ids:
+                        if i in punct_r_strip_candidate_token_ids:
                             token_id_to_word_id[i] = i
+                            word_id_to_token_ids[i].append(i)
                             word_id_to_text[i] = document.tokens[i].text
             else:
                 pass
 
+        # 2022-12; we need to handle cases like '(CS)' where the word starts with a punct "("
+        # we actually do want this as a separate word. but under the above logic, we would've
+        # assigned "(" and "CS" the same word ID.
+        # this part identifies words that begin with punctuation, and splits them.
```

Comment:
this whole file is kind of complex and getting harder to decipher so the comments are super helpful! TY

Reply:
thxx; unfortunately the more I look at DWP, the more I think this rules-based approach is too brittle & needs to be replaced w/ a model. The rules are super hard to maintain now & I'm worry about how difficult it is for others to debug/contribute to this Predictor as-is. Warrants an investigating a more robust solution